architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,186 @@
+
+This document will not include designs for the proposed measures to each threat.
+
+## System Description
+
+The Controller component is where we keep our Kubernetes controllers, a validating webhook, and a few other important parts. We will dive into more details on the controller below.
+
+Note: If you would like to learn more about Kubernetes controllers, you can visit [this link](https://kubernetes.io/docs/concepts/architecture/controller/).
+
+### Architecture
+
+The Controller component consists of several key parts:
+
+- **Recipe and Deployment Controllers**: Custom Kubernetes controllers registered with the manager. Each controller is responsible for watching specific Kubernetes resources (Recipe and Radius Deployment) and reconciling their state.
+- **Recipe Validating Webhook**: [Used if TLS certificates are provided] This webhook is triggered by the Kubernetes API in case of a change in a Recipe resource. Kubernetes API reaches out to this webhook in case of a creation, an update, or a deletion of a Recipe. Webhook validates the action, if it is a valid action, and responds to the Kubernetes API. For more information about webhooks, refer to the [official documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+- **Health Checks**: Health checks are implemented to monitor the status and performance of the controllers. They ensure that the controllers are functioning correctly and can take corrective actions if any issues are detected.
+
+### Clients
+
+In this section we are going to be talking about different clients of the Controller component. Clients are systems that reach out to the Controller component to trigger an action. Here are the clients of the Controller component:
+
+1. **Kubernetes API Server**: The primary client that interacts with the controller. It sends events related to resource changes (e.g., creation, update, or deletion of a Recipe and/or Deployment) to the controller manager. The controller watches for these events and reconciles the state of the resources accordingly.
+1. **Validating Webhook Clients**: If webhooks are enabled and registered, clients that interact with the Kubernetes API server (e.g., `kubectl`, other controllers) can trigger webhook calls.
+1. **Health Check Probes**: Kubernetes itself can act as a client by performing health and readiness checks on the controller manager.
+1. **Metrics Scrapers**: If metrics are enabled, Prometheus or other monitoring tools can scrape metrics from the controller manager.
+
+## Data Flow
+
+### Diagram
+
+![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)
```

Comment:
Would it make sense to show the user interaction, i.e., which trust boundaries are crossed by the user and which component(s) they interact with?

Reply:
I can add the User Interaction in this diagram too but in the Controller's case users don't directly reach out to the Controller. They make a change to a resource and then Kubernetes API Server reaches out to the Controller.

Reply:
`kubectl` is already authenticating and authorizing the user so it is not in Radius' scope. They have already crossed the trust boundary.

Reply:
Add user interaction to the diagram.

Reply:
Updated the diagram.

Reply:
Should there be numbers on the arrows to indicate sequence of events? or is that in another place?